### PR TITLE
Project-based permissioning framework

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

> I was rebasing #5894 on top of #5971, but I did it the wrong way somehow, because the latter got merged in the process.

Framework for #4263.
This is chunk 1 out of 3 of Per-project access, containing the data model, permissioning of managing project membership, and some essential frontend for controlling this.
PR 2 will have the UI for adding users to projects, while PR 3 will introduce hard restriction on access to all endpoints based on effective level (which is currently only the case for the new explicit_memberships endpoint). This one is up for review.

The feature being added is a premium one (meaning: `ee` directory whenever possible + `AvailableFeature` checks) plus feature-flagged currently.

See https://github.com/PostHog/posthog.com/pull/1948 for feature documentation.

~~Screenshots of new UI:~~ (out of date, see comment at the bottom)

### ~~Switch in organization settings (only if Per-project access is available)~~
Zrzut ekranu 2021-09-15 o 16 28 56

### ~~Members list in project settings (only if Per-project access is available and enabled)~~
Zrzut ekranu 2021-09-15 o 16 27 55

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [x] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
	- The removal of `users` is not completely obvious, but it was an old and confusing field, which is completely safe to remove because instead of a _column_ it's a separate many-to-many bridge table instead
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

File: posthog/api/user.py [hunk 1]
```diff
@@ -40,6 +39,7 @@ class Meta:
         model = User
         fields = [
             "date_joined",
+            "id",
```
Comment (paolodamico) [inline posthog/api/user.py:42]: Maybe worth merging #5971 first to make it super explicit what we're looking at here

File: posthog/models/organization.py [hunk 2]
```diff
@@ -79,8 +79,9 @@ class PluginsAccessLevel(models.IntegerChoices):
     updated_at: models.DateTimeField = models.DateTimeField(auto_now=True)
     domain_whitelist: ArrayField = ArrayField(
         models.CharField(max_length=256, blank=False), blank=True, default=list
-    )  # used to allow self-serve account creation based on social login (#5111)
+    )  # Used to allow/block self-serve account creation based on social login (#5111)
     setup_section_2_completed: models.BooleanField = models.BooleanField(default=True)  # Onboarding (#2822)
+    per_project_access: models.BooleanField = models.BooleanField(default=False)
```
Comment (paolodamico) [inline posthog/models/organization.py:84]: Suggestion: `project_based_permissioning` (or something around that?)

File: posthog/models/team.py [hunk 3]
```diff
@@ -114,10 +113,6 @@ class Team(UUIDClassicModel):
     plugins_opt_in: models.BooleanField = models.BooleanField(default=False)
     # DEPRECATED, DISUSED: replaced with env variable OPT_OUT_CAPTURE and User.anonymized_data
     opt_out_capture: models.BooleanField = models.BooleanField(default=False)
-    # DEPRECATED, DISUSED: now managing access in an Organization-centric way
-    users: models.ManyToManyField = models.ManyToManyField(
```
Comment (paolodamico) [inline posthog/models/team.py:118]: Should we document this in the API docs?

File: posthog/test/test_organization_model.py [hunk 4]
```diff
@@ -61,46 +55,3 @@ def test_plugins_are_not_preinstalled_on_cloud(self, mock_get):
 
         self.assertEqual(Plugin.objects.filter(organization=new_org, is_preinstalled=True).count(), 0)
         self.assertEqual(mock_get.call_count, 0)
-
-    @pytest.mark.ee
-    @patch("posthog.models.organization.License.PLANS", {"enterprise": ["whatever"]})
-    @patch("ee.models.license.requests.post")
-    def test_feature_available_self_hosted_has_license(self, patch_post):
```
Comment (paolodamico) [inline posthog/test/test_organization_model.py:68]: Feedback: would've been helpful if you added a review comment here specifying you copied this verbatim to the test class in the `ee` folder

File: posthog/migrations/0168_per_project_access.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("posthog", "0167_feature_flag_override"),
+    ]
+
+    operations = [
+        migrations.RemoveField(model_name="team", name="users",),
```
Comment (paolodamico) [inline posthog/migrations/0168_per_project_access.py:13]: IIRC we can't remove fields without pushing a full v2.0 version? But unsure about this. CC @mariusandra @timgl ?

File: posthog/api/organization.py [hunk 6]
```diff
@@ -99,8 +102,18 @@ def get_membership_level(self, organization: Organization) -> Optional[Organizat
         ).first()
         return membership.level if membership is not None else None
 
-    def get_setup(self, instance: Organization) -> Dict[str, Union[bool, int, str, None]]:
+    def get_only_allowed_team_ids(self, organization: Organization) -> Optional[List[int]]:
+        if not settings.EE_AVAILABLE or not organization.per_project_access:
+            return None  # All projects are allowed if Per-project access is disabled or unavailable
+        # If Per-project access is enabled, we need to check which projects the current user is a member of
```
Comment (paolodamico) [inline posthog/api/organization.py:108]: What if the user is an org admin/owner?

File: posthog/api/organization.py [hunk 7]
```diff
@@ -80,6 +81,8 @@ class Meta:
             "available_features",
             "domain_whitelist",
             "is_member_join_email_enabled",
+            "only_allowed_team_ids",
```
Comment (paolodamico) [inline posthog/api/organization.py:84]: I think this might be better computed by using the effective level for each team inside the `teams` attribute. It's less bloated, one single source of truth, and less queries and less computation in a critical endpoint.

File: posthog/api/team.py [hunk 8]
```diff
@@ -64,7 +69,28 @@ class Meta:
...
             "updated_at",
             "ingested_event",
+            "effective_membership_level",
+        )
+
+    def get_effective_membership_level(self, team: Team) -> Optional[int]:
+        requesting_parent_membership: OrganizationMembership = OrganizationMembership.objects.only("id", "level").get(
+            organization_id=team.organization_id, user=self.context["request"].user
         )
+        if not settings.EE_AVAILABLE:
```
Comment (paolodamico) [inline posthog/api/team.py:79]: Should you check here too that project-based permissioning is enabled for this project?

File: posthog/api/team.py [hunk 9]
```diff
@@ -64,7 +69,28 @@ class Meta:
...
+            # Per-project access not available
+            return requesting_parent_membership.level
+        from ee.api.explicit_team_member import ExplicitTeamMembership
+
+        try:
+            team_membership = ExplicitTeamMembership.objects.only("level").get(
+                team=team, parent_membership=requesting_parent_membership
+            )
+            return max(requesting_parent_membership.level, team_membership.level)
```
Comment (paolodamico) [inline posthog/api/team.py:88]: Repeated logic from a model, we should have this in a single place to avoid confusion or tech debt

File: posthog/api/test/test_organization.py [hunk 10]
```diff
@@ -69,17 +69,22 @@ def test_cant_create_organization_without_valid_license_on_self_hosted(self):
...
+                f"/api/organizations/{self.organization.id}", {"is_member_join_email_enabled": False}
+            )
+            response_per_project_access = self.client.patch(
+                f"/api/organizations/{self.organization.id}", {"per_project_access": False}
+            )
+            self.assertEqual(response_rename.status_code, status.HTTP_200_OK)
+            self.assertEqual(response_email.status_code, status.HTTP_200_OK)
+            self.assertEqual(response_per_project_access.status_code, status.HTTP_200_OK)
+            self.organization.refresh_from_db()
+            self.assertEqual(self.organization.name, "QWERTY")
```
Comment (paolodamico) [inline posthog/api/test/test_organization.py:87]: Check `is_member_join_email_enabled` & `per_project_access` in the model are updated too.

File: posthog/api/test/test_organization.py [hunk 11]
```diff
@@ -69,17 +69,22 @@ def test_cant_create_organization_without_valid_license_on_self_hosted(self):
...
-        )
-        self.assertEqual(response_rename.status_code, status.HTTP_200_OK)
-        self.assertEqual(response_email.status_code, status.HTTP_200_OK)
-        self.organization.refresh_from_db()
-        self.assertEqual(self.organization.name, "QWERTY")
+    def test_update_organization_if_admin_or_owner(self):
+        for level in (OrganizationMembership.Level.ADMIN, OrganizationMembership.Level.OWNER):
+            self.organization_membership.level = level
+            self.organization_membership.save()
+            response_rename = self.client.patch(f"/api/organizations/{self.organization.id}", {"name": "QWERTY"})
```
Comment (paolodamico) [inline posthog/api/test/test_organization.py:76]: To make it relevant, you should change this attribute (and ideally the other two below) based on the level, otherwise it could technically be that only one of them is updating correctly

File: posthog/api/routing.py [hunk 12]
```diff
@@ -47,11 +48,15 @@ def team(self) -> Team:
...
 
+        user = cast(User, self.request.user)
         if self.legacy_team_compatibility:
-            team = self.request.user.team
+            team = user.team
             assert team is not None
             return team
-        return Team.objects.get(id=self.team_id)
+        try:
+            return Team.objects.get(id=self.team_id)
```
Comment (paolodamico) [inline posthog/api/routing.py:57]: Why are we doing this extra check when you could just return `user.team`, FK will already protect that the project exists.

File: ee/api/test/test_team_memberships.py [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+
+        new_user: User = User.objects.create_and_join(self.organization, "rookie@posthog.com", None)
+
+        self.assertEqual(self.team.explicit_memberships.count(), 0)
+
+        response = self.client.post("/api/projects/@current/explicit_members/", {"user_uuid": new_user.uuid})
+        response_data = response.json()
+
+        self.assertDictEqual(
+            self.permission_denied_response("You don't have sufficient permissions in this project."), response_data
```
Comment (paolodamico) [inline ee/api/test/test_team_memberships.py:64]: "[...] for this project or organization."

File: ee/api/test/test_team_memberships.py [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+        self.assertEqual(self.team.explicit_memberships.count(), 0)
+
+    def test_add_yourself_as_org_member_forbidden(self):
+        self.organization_membership.level = OrganizationMembership.Level.MEMBER
```
Comment (paolodamico) [inline ee/api/test/test_team_memberships.py:71]: Maybe better to test as `ADMIN` or `OWNER` as permissions are not the scope of this test case.

File: ee/api/test/test_team_memberships.py [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+        response_data = response.json()
+
+        self.assertDictEqual(
+            self.permission_denied_response("You don't have sufficient permissions in this project."), response_data
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+        self.assertEqual(self.team.explicit_memberships.count(), 0)
+
+    def test_add_yourself_as_org_admin_forbidden(self):
```
Comment (paolodamico) [inline ee/api/test/test_team_memberships.py:86]: I would just keep the test case above

File: ee/api/test/test_team_memberships.py [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+
+        self.assertDictContainsSubset(
+            {"effective_level": ExplicitTeamMembership.Level.MEMBER, "level": ExplicitTeamMembership.Level.MEMBER,},
+            response_data,
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        self.assertEqual(self.team.explicit_memberships.count(), 2)
+
+    def test_add_member_as_org_admin_and_project_member_allowed(self):
```
Comment (paolodamico) [inline ee/api/test/test_team_memberships.py:145]: Looking at this test case, should we maybe restrict that admins and owners can't be added as project members? This way we avoid weird edge cases for when the user's org permissions changes. It also avoids extra handling such as `test_add_admin_as_project_member_forbidden`

File: ee/api/test/test_team_memberships.py [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+        response = self.client.post("/api/projects/@current/explicit_members/", {"user_uuid": new_user.uuid})
+        response_data = response.json()
+
+        self.assertDictContainsSubset(
+            {"effective_level": ExplicitTeamMembership.Level.MEMBER, "level": ExplicitTeamMembership.Level.MEMBER,},
+            response_data,
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_add_admin_as_org_admin_allowed(self):
```
Comment (paolodamico) [inline ee/api/test/test_team_memberships.py:163]: Same comment as above.

File: ee/api/test/test_team_memberships.py [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+        self.organization_membership.save()
+        _, new_team, new_user = User.objects.bootstrap("Acme", "mallory@acme.com", None)
+
+        response = self.client.post(f"/api/projects/{new_team.id}/explicit_members/", {"user_uuid": new_user.uuid,})
+        response_data = response.json()
+
+        self.assertDictEqual(self.not_found_response("Project not found."), response_data)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_add_member_to_nonexistent_project_forbidden(self):
```
Comment (paolodamico) [inline ee/api/test/test_team_memberships.py:249]: Another test case could be not being able to add a non-member of your org to a project to which you would have permissions.

File: frontend/src/types.ts [hunk 19]
```diff
@@ -86,25 +85,53 @@ export interface OrganizationBasicType {
...
     setup: SetupState
     setup_section_2_completed: boolean
+    per_project_access: boolean
     plugins_access_level: PluginsAccessLevel
     teams: TeamBasicType[] | null
     available_features: AvailableFeature[]
     domain_whitelist: string[]
     is_member_join_email_enabled: boolean
+    membership_level: OrganizationMembershipLevel | null
+    only_allowed_team_ids: TeamType['id'][] | null
```
Comment (paolodamico) [inline frontend/src/types.ts:98]: See comment on backend section

File: frontend/src/scenes/saved-insights/SavedInsights.tsx [hunk 20]
```diff
@@ -68,7 +69,8 @@ export function SavedInsights(): JSX.Element {
     } = useValues(savedInsightsLogic)
     const { dashboards } = useValues(dashboardsModel)
     const { hasDashboardCollaboration } = useValues(organizationLogic)
-    const { members } = useValues(membersLogic)
+    const { currentTeam } = useValues(teamLogic)
+    const { members } = useValues(membersLogic({ team: currentTeam }))
```
Comment (paolodamico) [inline frontend/src/scenes/saved-insights/SavedInsights.tsx:73]: Why are you passing this prop? `membersLogic` never reads/expects it.

File: frontend/src/scenes/events/definitions/DefinitionOwnerDropdown.tsx [hunk 21]
```diff
@@ -5,9 +5,11 @@ import { membersLogic } from 'scenes/organization/Settings/membersLogic'
 import { UserBasicType } from '~/types'
 import { definitionDrawerLogic } from './definitionDrawerLogic'
 import { Owner } from '../Owner'
+import { teamLogic } from '../../teamLogic'
 
 export function DefinitionOwnerDropdown({ owner }: { owner: UserBasicType | null }): JSX.Element {
-    const { members } = useValues(membersLogic)
+    const { currentTeam } = useValues(teamLogic)
+    const { members } = useValues(membersLogic({ team: currentTeam }))
```
Comment (paolodamico) [inline frontend/src/scenes/events/definitions/DefinitionOwnerDropdown.tsx:12]: Why are you passing this prop? `membersLogic` never reads/expects it.

File: frontend/src/scenes/organization/Settings/index.tsx [hunk 22]
```diff
@@ -161,6 +165,15 @@ export function OrganizationSettings({ user }: { user: UserType }): JSX.Element
                 <Divider />
                 <Members user={user} />
                 <Divider />
+                {featureFlags[FEATURE_FLAGS.PER_PROJECT_ACCESS] && (
+                    <>
+                        <RestrictedArea
+                            Component={Permissioning}
```
Comment (paolodamico) [inline frontend/src/scenes/organization/Settings/index.tsx:171]: I was actually extremely confused when I saw this. I thought you enabled project-based permissioning at the project-level, not org level. That way you can have "default" projects that everyone has access to, and then you can create private projects if you want to. Seems a tad more intuitive. Thoughts?

**Review by paolodamico — CHANGES_REQUESTED** [review]
Alright, did an initial pass of this. Didn't review the actual UI code yet, as the other feedback might change that. I also haven't actually tested this in the app.

Feedback: I think this should've been a PR for the backend only, and then a separate one for frontend changes.

File: frontend/src/scenes/organization/Settings/index.tsx (hunk 22 above)
Comment (Twixes) [reply frontend/src/scenes/organization/Settings/index.tsx:171]: Hmm, I thought we agreed upon this in the approved documentation PR, which outlines the mechanism implemented.
I see your point though, I think both are intuitive but entirely project-based does sound a bit more flexible. I can rework this if you think it's worth it?

File: ee/api/test/test_team_memberships.py (hunk 13 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:64]: Current wording seems the clearest to me, in that elevated organization permissions aren't actually needed – having elevated project-specific perms is enough.

File: ee/api/test/test_team_memberships.py (hunk 14 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:71]: They are in fact the scope of this test case. This tests that prohibition on the basis of insufficient permissions takes precedence over prohibition on the basis of the subject being the user themselves. The test case for admins (`test_add_yourself_as_org_member_forbidden`) below tests that prohibition on the basis of the subject being the user themselves works too.

File: ee/api/test/test_team_memberships.py (hunk 15 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:86]: See response above.

File: ee/api/test/test_team_memberships.py (hunk 16 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:145]: IMO it doesn't make sense to normalize this further, as then we easily end up in a situation where a previous plain organization member gets added to a project and elevated org-wide later on, in which case checks project-member-addition-time are moot.

File: ee/api/test/test_team_memberships.py (hunk 17 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:163]: See response above.

File: ee/api/test/test_team_memberships.py (hunk 18 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:249]: For sure

File: ee/models/explicit_team_membership.py [hunk 23]
```diff
@@ -0,0 +1,45 @@
+from django.db import models
+
+from posthog.models.utils import UUIDModel, sane_repr
+
+
+class ExplicitTeamMembership(UUIDModel):
```
[the remark this answers is not in the document]
Comment (paolodamico) [reply ee/models/explicit_team_membership.py:6]: Makes sense! We can then make one swift change when the time comes.

File: frontend/src/scenes/organization/Settings/index.tsx (hunk 22 above)
Comment (paolodamico) [reply frontend/src/scenes/organization/Settings/index.tsx:171]: Sorry I completely misunderstood when reading the docs, got confused on the way we "turn on" project-based permissioning. This will be complex to change in the future so we should ship with whatever makes the most sense, which I'm still unsure about. 

Would love to get @clarkus & @marcushyett-ph (as he's been chatting with customers more frequently on this) input on this too. TLDR; you can now have project-based permissions in addition to org-based permissions. Question is, should we enable this at the org-level (wherein if enabled, every project will be private by default for non-admins) or at the project-level (where in every project will be public to all org members by default unless you turn this setting on for the specific project).

File: posthog/models/organization.py (hunk 2 above)
Comment (paolodamico) [reply posthog/models/organization.py:84]: We can always change the user-facing descriptions / naming, agreed it doesn't make change to change internally for now.

File: posthog/migrations/0168_per_project_access.py (hunk 5 above)
Comment (paolodamico) [reply posthog/migrations/0168_per_project_access.py:13]: Perfect! Maybe worth adding that text snippet to the actual migration just for future reference?

File: posthog/api/routing.py (hunk 12 above)
Comment (paolodamico) [reply posthog/api/routing.py:57]: Got it, makes sense!

File: ee/api/test/test_team_memberships.py (hunk 16 above)
Comment (paolodamico) [reply ee/api/test/test_team_memberships.py:145]: My concern is from a UX-standpoint, it's confusing that if you're an admin/owner you can be added to projects only under certain conditions. It's simpler to communicate that you can never be added to projects if you're an admin because you have access to everything.

**Comment by paolodamico** [comment]
@Twixes I'll wait for the resolution on org-level vs. project-level setting and then finish the review, sounds good?

File: frontend/src/scenes/organization/Settings/index.tsx (hunk 22 above)
Comment (marcushyett-ph) [reply frontend/src/scenes/organization/Settings/index.tsx:171]: I think we need to support both starting points:

e.g. If you've been using PostHog for a while and everyone expects to have access to everything, but you turn on this feature - you'd expect that people can still access everything until specific permissions have been configured

e.g. If they're setting up PostHog for the first time and have no users, they would probably want to grant permissions only to specific people on specific projects.

Comment (paolodamico) [reply frontend/src/scenes/organization/Settings/index.tsx:171]: I'm actually a bit against this, I think this will be confusing for users and harder to manage. I believe in any of the options both existing and new users will have the option to treat things as they want.

Comment (Twixes) [reply frontend/src/scenes/organization/Settings/index.tsx:171]: Given these examples, I think option 1 (what Paolo suggests) seems to serve both users really well (I'm also imagining that when creating new projects, they would just have a checkbox "restrict default access"), so I believe it makes sense to implement, though it will make the feature spill over a bit more into the next sprint.

File: ee/api/test/test_team_memberships.py (hunk 16 above)
Comment (Twixes) [reply ee/api/test/test_team_memberships.py:145]: I totally agree UX-wise! The UI won't allow adding any admin/owner to projects. But in the backend I worry the code would just get more complex kind of for no significant gain.
BTW `test_add_admin_as_project_member_forbidden` definitely can't be avoided – the name definitely is a bit ambiguous as I had to keep it pretty concise, but what it tests is that, being a plain project member, you can't add another organization member (who doesn't belong to the project yet) as a project admin.

File: frontend/src/scenes/organization/Settings/index.tsx (hunk 22 above)
Comment (paolodamico) [reply frontend/src/scenes/organization/Settings/index.tsx:171]: I think it's fine to have a bit of spill unto this next sprint as it's very related to what you're working on anyways, but you'll have better context here

File: ee/api/test/test_team_memberships.py (hunk 16 above)
Comment (paolodamico) [reply ee/api/test/test_team_memberships.py:145]: Got it! Well if we don't allow it in the frontend, I'm totally fine with this!

**Review by paolodamico — CHANGES_REQUESTED** [review]
Did another pass, but I think it's worth waiting until we do the change to be project-first. One frontend thing I noticed is that I see a project disabled yet: a) `per_project_access` is `False`, b) the feature flag is off, c) this is the default project, d) it's the project I'm currently at.

**Comment by Twixes** [comment]
Sure, will make changes now. Also, I have to say I found iterating on the backend and the frontend at the same time quite useful, so I disagree a bit that this should be an all-backend PR. But the project switcher definitely needs some more polish and is more of an addon, so I extracted it to its own PR: https://github.com/PostHog/posthog/pull/6027

**Comment by Twixes** [comment]
After reorientation this PR adds the following UI (instead of what's shown in PR description):

### Project-based permissioning on
On

### Project-based permissioning off
Off

File: posthog/models/organization.py (hunk 2 above)
Comment (Twixes) [reply posthog/models/organization.py]: Alrighty, I actually did end up changing this to `project_based_permissioning` as I had to rework fields anyway to make this completely project-based.

File: posthog/models/organization.py [hunk 24]
```diff
@@ -79,7 +79,7 @@ class PluginsAccessLevel(models.IntegerChoices):
     updated_at: models.DateTimeField = models.DateTimeField(auto_now=True)
     domain_whitelist: ArrayField = ArrayField(
         models.CharField(max_length=256, blank=False), blank=True, default=list
-    )  # used to allow self-serve account creation based on social login (#5111)
+    )  # Used to allow/block self-serve account creation based on social login (#5111)
```
Comment (paolodamico) [inline posthog/models/organization.py:82]: Technically this doesn't block self-serve account creation. If you log in with SSO and your base domain does not match this list you can still use an invite (i.e. is completely disregarded)

**Review by paolodamico — APPROVED** [review]
alright did another quick code review pass, and adjusted the copy a bit based on the doc changes.

I think it's worth merging all upstream PRs and getting this into `master` as it's getting quite unwieldy. My only ask is please **wait to merge until we branch off the 1.28.1 patch release** (a couple of hours at most)

File: frontend/src/scenes/project/Settings/AccessControl.tsx [hunk 25]
```diff
@@ -6,25 +6,35 @@ import { useActions, useValues } from 'kea'
...
+    const projectPermissioningEnabled =
+        currentOrganization?.available_features.includes(AvailableFeature.PROJECT_BASED_PERMISSIONING) &&
+        currentTeam?.project_based_permissioning
+
     return (
         <div>
-            <h2 className="subtitle" id="access-control">
-                Access Control
+            <h2 className="subtitle" id="privacy">
+                Privacy settings
```
Comment (Twixes) [inline frontend/src/scenes/project/Settings/AccessControl.tsx:24]: I'd prefer to stick to "Access Control". "Privacy" can be confused for user privacy (GDPR and such), while "Access Control" clearly refers to project access.

File: frontend/src/scenes/saved-insights/SavedInsights.tsx [hunk 26]
```diff
@@ -32,6 +32,7 @@ import { dashboardsModel } from '~/models/dashboardsModel'
 import { DateFilter } from 'lib/components/DateFilter/DateFilter'
 import '../insights/InsightHistoryPanel/InsightHistoryPanel.scss'
 import dayjs from 'dayjs'
+
```
Comment (mariusandra) [inline frontend/src/scenes/saved-insights/SavedInsights.tsx:35]: 🤔

**Review by mariusandra — APPROVED** [review]
Went through it all... and LGTM!

I did find some inline nits (how could I not), and then UX wise I'm not really sure what's enabled when "Project Based Permissioning" is successfully updated. All I did was to make my project private.



Other than that, really excited to get this in.

File: frontend/src/scenes/project/Settings/TeamMembers.tsx [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+            sorter: (a, b) => a.joined_at.localeCompare(b.joined_at),
+            defaultSortOrder: 'ascend',
+        },
+        /*{
+            key: 'actions',
+            align: 'center',
+            render: function ActionsRender(_, member) {
+                return ActionsComponent(member)
+            },
+        },*/
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/project/Settings/TeamMembers.tsx:141]: Yup: https://github.com/PostHog/posthog/pull/6067

File: ee/api/explicit_team_member.py [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+        except OrganizationMembership.DoesNotExist:
+            return True  # This will be handled as a 404 too
+        if requesting_team_membership is None:
+            return False
+        minimum_level = (
+            ExplicitTeamMembership.Level.MEMBER
+            if request.method in SAFE_METHODS
+            else ExplicitTeamMembership.Level.ADMIN
+        )
+        return requesting_team_membership.effective_level >= minimum_level
```
[the remark this answers is not in the document]
Comment (Twixes) [reply ee/api/explicit_team_member.py:60]: Not sure I get this. If you get True from `requesting_team_membership.effective_level >= minimum_level`, you get a 2xx.

[the remark this answers is not in the document]
Comment (mariusandra) [reply ee/api/explicit_team_member.py:60]: ah, I get how this works now :)
